sdks/go/pkg/beam/external_test.go

```diff
@@ -0,0 +1,104 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package beam
+
+import (
+	"bytes"
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/typex"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/util/reflectx"
+	"github.com/apache/beam/sdks/go/pkg/beam/options/jobopts"
+)
+
+func TestExternalTagged(t *testing.T) {
+	urn := "my::test::urn"
+	payload := []byte("expected_payload")
+	inTag := "inTag"
+	outTag := "outTag"
+
+	p := NewPipeline()
+	s := p.Root()
+
+	col := Impulse(s)
+	ft := typex.New(reflectx.ByteSlice)
+	outMap := ExternalTagged(s, urn, payload, map[string]PCollection{inTag: col}, map[string]FullType{outTag: ft}, true)
+
+	// Validate the in-construction PCollection types for ExternalTagged.
+	if len(outMap) != 1 {
+		t.Fatalf("ExternalTagged = %v, want 1 output with tag %q", outMap, outTag)
+	}
+	out, ok := outMap[outTag]
+	if !ok {
+		t.Fatalf("ExternalTagged = %v, want 1 output with tag %q", outMap, outTag)
+	}
+	if !typex.IsEqual(out.Type(), ft) {
+		t.Fatalf("ExternalTagged()[%q].Type() = %v, want %v", outTag, out.Type(), ft)
+	}
+
+	// Validate the post-construction pipeline protocol buffer contents.
+
+	ctx := context.Background()
+	envUrn := jobopts.GetEnvironmentUrn(ctx)
+	getEnvCfg := jobopts.GetEnvironmentConfig
+	environment, err := graphx.CreateEnvironment(ctx, envUrn, getEnvCfg)
+	if err != nil {
+		t.Fatalf("Couldn't create environment build: %v", err)
+	}
+
+	edges, _, err := p.Build()
+	if err != nil {
+		t.Fatalf("Pipeline couldn't build: %v", err)
+	}
+	pb, err := graphx.Marshal(edges, &graphx.Options{Environment: environment})
+	if err != nil {
+		t.Fatalf("Couldn't graphx.Marshal edges: %v", err)
+	}
+	components := pb.GetComponents()
+	transforms := components.GetTransforms()
+
+	foundExternalTagged := false
+	for _, transform := range transforms {
+		spec := transform.GetSpec()
```

Comment:
This loop seems a bit strangely structured. From what I can tell it's just searching for the external transform by URN and then doing some checks, right? You could just put all the checks in the `if strings.Contains` statement and do all the validation at once, or if you want to avoid indentations, then reverse it by starting the loop with `if !strings.Contains { continue; }`.

Reply:
Ooo good catch. Thanks!